### ISSUE TITLE
Real-time vector art rendering (e.g. SVG)

### ISSUE THREAD

**Issue by projectitis** [godotengine/godot-proposals#2924]
### Describe the project you are working on
Game development using vector art work. I'm a long-time (former) Flash/AIR/ActionScript3 developer, almost exclusively using vector art assets. 

### Describe the problem or limitation you are having in your project
Although vector art can be imported into Godot, it is rasterized on import and is a bitmap from that point and is not rendered in real time. Real-time vector art rendering has several benefits over bitmap art, which include:
* Scalable (and rotatable, transformable etc) without quality loss
* Much smaller file size
* Transformable in real time, and transformations are not destructive (e.g. can be reversed)
* Many vector asset types can be managed in source control as they are essentially text based

Yes, there are negatives too - the main one being much slower to render than bitmaps, and lack of GPU acceleration. However, many game developers are willing to make this trade-off for the benefits that vector art can provide,

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
I propose a new Node2D descendent which supports real-time vector rendering using [Blend2D](https://blend2d.com).

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams
The mechanism, if this is the way recommended by the core devs, would be basically to use the [custom drawing feature of CanvasItem](https://docs.godotengine.org/en/stable/tutorials/2d/custom_drawing_in_2d.html) to update the node (draw a new texture) if the vector updates. Ideally an SVG asset would be assigned to the node.  Eventually, animation could be supported via multiple SVG images, or by deforming or transforming the vector geometry directly.

I have already forked Godot and have started this work. But any recommendations, tips, feedback from the core devs and other Godot contributors would be very welcome.

### If this enhancement will not be used often, can it be worked around with a few lines of script?
The work-around is not to use vector art, which is really not a workaround :)

### Is there a reason why this should be core and not an add-on in the asset library?
If available, it should be a core feature and maintained as part of the product.

**Comment by pycbouh** [comment]
> However, many game developers are willing to make this trade-off for the benefits that vector art can provide,

Well, if you're making this claim it would be nice to know where you're getting this metric from. 👀 

If this to be implemented, you have to make sure that users aren't confused why their vector assets suddenly cause way worse performance.

**Comment by Calinou** [comment]
https://github.com/godotengine/godot/pull/44772 implements this for SVGs using the same rendering technology as MSDF fonts.

**Comment by fire** [comment]
My approach for doing real-time vector art rendering is step by step via https://github.com/godotengine/godot/pull/49645.

Stage 1: Implement thorvg. Notice the render commands in the documentation.
Stage 2: Evaluate either rive https://github.com/rive-app/rive-tizen or a future port of lottie vector animations from the same Samsung team as thorvg. See thorvg's roadmap https://github.com/Samsung/thorvg/wiki/21'-Development-Milestone

**Comment by projectitis** [comment]
Thanks for the comments and considerations everyone.
@fire - those prototypes and other (animated) SVG projects are very interesting, especially Rive.

**Comment by projectitis** [comment]
I finally have a working c++ version playing [Rive](https://rive.app/) files using [Skia](https://github.com/google/skia) as the rendering engine. Not in Godot, just a standalone app at present.

Rive is a vector format for animation (much more than just animation, actually) which would be an amazing addition to Godot.

I hope that @fire or someone else is able to implement a thorvg renderer in time! (see [skia renderer](https://github.com/rive-app/rive-cpp/tree/master/skia/renderer) for reference).

**Comment by fire** [comment]
https://github.com/godotengine/godot/pull/49645 is currently merged into master for ThorVG vector graphics support.

Currently the svg module is using the standard importer design, but there's no real reason to be only offline importable.

Here's an early preview on SVG to Mesh. Not expecting to finish SVG to mesh before Godot 4.0 release though.

![image](https://user-images.githubusercontent.com/32321/154733311-4c3ee071-5943-4bfe-bd05-21295f0e29a1.png)

![image](https://user-images.githubusercontent.com/32321/154733330-eed4b9cb-0548-4569-9e3e-66620f4e61b7.png)

**Comment by Sosasees** [comment]
Resolution-independent graphics are Really useful as we're already in the way of having every screen be a HiDPI screen.

Every mobile phone already has a very dense screen with at least 144dpi, and even many of the mid-range computer monitors are already denser than 72dpi even if they don't quite reach true HiDPI yet (e.g. my 1920×1080px monitor is 80.68dpi, which is rounded up to 96dpi by Debian Linux or by Godot, i'm not sure which).

My current workaround as of Godot 3.4.4 is to use Distance Field shaders, which is my only use of custom shaders so far.

**Comment by fire** [comment]
These are currently the three ways to use vector graphics in Godot Engine 4.

1. Raster vector graphics to a bitmap through thorvg importing.
2. Save the vector graphics to a mesh https://github.com/V-Sekai/godot-svg-mesh using Tove2D.
3. Render a font (multichannel) signed distance field.

Missing:

1. Render 2d vectors graphics as a signed distance field and use in Godot Engine 2D.
2. Save a 2d vector graphic to a mesh using Thorvg instead of Tove2D. https://github.com/V-Sekai/godot-svg-mesh

**Comment by Giwayume** [comment]
https://github.com/Giwayume/godot-svg

Here is a proof of concept demonstrating that bezier curves can be drawn in shader code, using a SVG to mesh conversion approach. Curves are not tessellated, they are perfect at every resolution.

**Comment by Sosasees** [comment]
@Giwayume i can't check your **g**odot project right now, but this sounds great

in past projects i have recreated some vector shapes as shaders to get infinite smoothness, because your proof-of-concept'd feature isn't in **g**odot.
(i have only made circle and squircle as shaders, because the other shapes i needed so far were just polygons, that i could create more easily but still lossless with Polygon2D nodes)

if this feature gets added, i can get these smooth shapes without shader code but from just importing my SVG files

**Comment by Giwayume** [comment]
This project takes a generalized approach. The technique can theoretically render any SVG shape, current bugs aside.

**Comment by SlugFiller** [comment]
A year ago, I made a realtime vector+SVG importer plugin for Godot 3. It's not complete, but it's never going to be complete, so I just posted it [here](https://github.com/SlugFiller/godot-vector2d). It supports shapes, strokes (via stroke to fill conversion), winding rules, and basic gradients. I was also planning to implement mesh gradients, but never got around to it.

Underneath, it works purely as a shader. It makes an approximation of the vector's bounding box, draws a rectangular quad around it, and uses a tracing algorithm in the fragment shader. I did hit quite a few difficulties:

1. Functions in a shader are not really functions. I couldn't use regular recursion, so I had to use a virtual stack instead.
2. In fact, I couldn't use functions at all, since I can't pass the virtual stack into a function, nor have two functions that do not call each other share the same virtual stack. The result is that the entire shader is one large function, with a lot of code repetition. This is a flaw in the shader language. If functions operate fundamentally as macros, it would have been better to have actual macros, so they can inherit stack variables from the parent. Alternately, just allow global variables, since all stack variables are technically implemented as global.
3. Lack of uniform arrays made it difficult to pass data to the shader. Using a texture has limitations, since floating point values are clamped to 0.0-1.0 range, so I can't pass arbitrary curve coordinates. I had to create my own floating point encoding that projects the real plane into the 0.0-1.0 range. Obviously, this carries an accuracy cost, but nothing visible for most use cases. Addition of uniform arrays in Godot 4 could fix this.
4. When a path is made of many segments, doing an iteration over all segments for every pixel was too slow, and Godot would occasionally crash. I alleviated the problem by creating a binary bounding box tree, which allows very quickly eliminating large portions of the shape, as well as getting a quick decisions on pixels that are not inside the bounding box of any segment. It can still slightly choke on complex shapes. If I could do multiple passes, I could first draw individual paths into a crossing buffer, then do scanline rendering using that buffer as input. But Viewports aren't able to act as integer buffers, nor do they allow accumulating data. And unfortunately, GDScript is too slow to do this in a CPU buffer, and also, isn't able to run every frame, and makes it difficult to get the current viewport matrix. In general, the whole thing could be much faster, if I could first do a "per y" calculation, and then use the results in a "per x" fragment. Maybe it's possible with computer shaders, not sure.
5. Dealing with viewport matrices in general was also difficult. I had to use varying to pass things like WORLD_MATRIX from the vertex shader to the fragment shader. Doing a projection of the vector in the fragment shader is a lot easier than trying to reverse project the FRAGCOORD, which also has its own issues.
6. I wanted to added feather measured in screen coordinates, rather than object coordinates, since that would allow me to get easy anti-aliasing by doing 0.5 pixels feather. However, it's not possible to do calculates based on screen coordinates in GDScript. What that meant was that I had to calculate the drawing bbox in the vertex shader. I simply passed a "sufficiently large" rectangle (so it wouldn't be culled for being out of screen) as a drawing command, the shape's bounding box in its local coordinates using uniforms, then used UV to figure out which point I am converting to turn the passed rect into something that actually wraps the shape and its pixel-sized feather.
7. A lack of a native dirty/flush mechanism in Godot, made it hard to get the vector objects to work with something like animation onion skinning while remaining efficient. For example, I use GDScript to do stroke to fill, but I don't want it to run for every point that is moved in the vector in a tween animation. I ultimately made a hack using the fact that a Spatial has special handling for its NOTIFICATION_TRANSFORM_CHANGED that acts like a dirty/flush. I made a request to have a non-hacky option in #2598.
8. Godot 3 has no clip shapes, which are a core SVG feature. Godot 4, thankfully, fixes this with Clip Children: Clip only.
9. Unrelated to the vector rendering itself, as I was making an editor for the vector shapes, I noticed a large gap between methods available on the C++ side, and methods available on the GDScript side. I opened a couple of suggestions where I hit roadblocks. And yes, I have an editor for the shapes, because I believe reliance on imports ignores the gap between what a graphics format that can be pre-processed for render should be capable of doing, and what a realtime interactive engine should be capable of doing.

Incidentally, vectors can actually give an advantage over extremely high resolution sprites, due to taking up less memory, especially on mobile platform. But I did test this using a browser engine with native vector support, not using Godot.

Overall, new features in Godot 4 might make it easier to implement an efficient vector shape renderer. But a native vector shape object, even if it's only shape fill (with winding rule support), that runs in the render pipeline, can save a significant amount of effort. Implementing gradients, stroker, shape tweens, etc, on top of that, is much easier.

So, IMO, import formats should be left to plugins, but a basic bezier curve shape fill with infinite smoothness and builtin editor should still be implemented in-engine. Preferably with quadric and arc support (although the prior can be accurately represented using cubic)

**Comment by Giwayume** [comment]
It's cool to see calculating the fill rule per-pixel in a shader is feasible for simple shapes, though the performance looks very poor.

On my laptop:
Max FPS: 600
SVG circle with fill: 180 FPS
SVG circle with stroke: 95 FPS
SVG circle with fill and stroke: 70 FPS

I actually can't get many SVGs to render in order to test more complex examples. I'd love to see improvements made to that project as theoretically it could be the best technique for rendering SVG animations that morph the path every frame.

**Comment by Sosasees** [comment]
i understand that the SVG format was made for websites, not games

SVG has many advanced features that help with making more detailed pictures at the cost of speed.
a 3D analogy: SVG is more similar to Blender files than glTF

you can already make vector graphics with Godot-native features like Polygon2D and shaders.
if SVG rendering can be made fast, it's great, but not as needed as some people might think

**Comment by Giwayume** [comment]
I think to SlugFilter's point, what people mostly want out of the engine is a way to render shapes with smooth bezier curves, that can be colored, textured, or used with a shader.

What I have presented is a concept where that shape is converted to a triangulated mesh, where each bezier command gets dedicated triangles and the calculation in the shader is simple linear algebra. This is very performant at runtime, but requires some work on the CPU beforehand to generate the mesh from the SVG path commands.

The approach SlugFilter shows could potentially be more performant for animating the `d` attribute of a `path` (animating transitions between curves), because the work on the CPU to convert a new path command to a texture that basically stores the path data is much less than all of the path solving work my project does to convert everything to a new mesh. Though, the mesh conversion approach I have shown is basically more performant for every single other scenario, and by a large margin.

I could envision Godot natively supporting a node that uses both techniques in an exchangeable manner, either automatically if it's feasible to detect that the developer is trying to animate the path, or through a boolean property on the node.

**Comment by SlugFiller** [comment]
@Sosasees The issue with both `Polygon2D` and `draw_*()` is that both only option for straight segments. (Well, there's a draw_arc, but it only draws a stroked arc, so you can't compose a filled shape from multiple arcs).

While a complex format like SVG isn't built for speed, something like CanvasRenderingContext2D::bezierCurveTo _is_ built for performance, and is used to effect in HTML5 2D games.

Even something as tiny as adding a `draw_filled_curve` method to `CanvasItem` (and a `canvas_item_add_filled_curve` to `VisualServer`, obviously) would allow filling in the rest using plugins and shaders.

In fact, I might prefer this minimal approach, because, despite the extra effort, I'm more likely to do well making and using my own curve editor, then rely on one designed by someone else.

@Giwayume The reason it's slow is because the fragment shader recalculates a lot of things a normal rasterizer would only need to calculate once. For example, if a bunch of path segments are completely above or below the current y, each fragment in the same y would need to make the same calculation to figure this out, even though it's the exact same calculation for all of them (I do winding counting by tracing a ray in the screen's x+ direction, so checks for the y value are largely uniform). Even though they do this in parallel, the GPU's number of cores is not infinite, and it takes resources that could be otherwise used for parallelizing for different y values.

As I've said above, if I had a "row" shader, that allowed me to make a bunch of calculations for a given y value, and then pass the results as an array to the fragment shader, it could vastly improve performance. This is how software rasterizers can fill curves very quickly. Hybrid rasterizers can do extract horizontal slices in CPU, then let the GPU simply draw horizontal line commands.

An alternative approach is to draw winding additively on a buffer, then rasterize each row left to right, summing up the winding as you go. The fragment shader's inability to run in any sort of sequence works against it, again, but this could be alleviated by storing the winding in a tree format, that would make it possible to look up the winding for a given fragment at `log(width)` complexity. The initial winding rasterizer can still be fairly performant if done in software. But not if that software is written in GDScript. Additionally, it's impossible to patch the render pipeline in Godot directly, so just figuring out the height of the screen, and the needed resolution for rasterization, and running it on every frame, would be impossible from GDScript.

Put more simply, my approach is NOT the best approach, it is the best approach given the limitations of working inside Godot 3. Browsers and programs like Inkscape, demonstrate rasterizing far more complex vectors, at significantly higher speed, due to using more effective rasterization methods that can't be implemented in pure GDScript or simple vertex-fragment shaders.

As for subdividing the shader rect, triangle culling wouldn't help with performance, because the fragment shader doesn't run for out-of-screen pixels anyway. If anything, it would just create extra effort for the GPU, as fragments near the seams have to run twice. The reason performance dies as you zoom in is simply because the shader is ran for more fragments.

It would be pretty easy to test in practice, though. Replace the draw_rect in fill.gd with a draw_mesh. The locations of the vertices in the source mesh don't really matter, only the UVs, since the vertex shader uses only the UVs to produce the actual vertex positions.

**Comment by Sosasees** [comment]
i like to make my game vector graphics as Godot shaders, mainly because i can add motion effects that i can't with SVG

i understand that many people don't want to write shaders, so i respect the effort of adding realtime SVG rendering to Godot — different from Godot's current SVG support, which justs imports SVG files as raster images (similar to 'Export PNG' in Inkscape)

**Comment by SlugFiller** [comment]
@Sosasees The issue is that shaders alone, in their current specs/limitations, are insufficient to implement generic and efficient bezier shape rendering. You can see that both my approach, and @Giwayume's approach, completely buckle under certain conditions. This is due to limitations of what the Godot shader specs expose, and how slow GDScript is for large iterations or recursion.

It's possible that Godot 4 exposes better APIs for making a more efficient shader. I haven't tested. It still couldn't possibly be as efficient as an engine-supported command.

Incidentally, if you only want infinitely smooth corners, and don't need beziers specifically, you could make a shader that turns a triangle into a filled arc, then make a smooth shape out of those. But it would be impossible to work with artists who expect standard tools, and very difficult to make any line effects and animations.

**Comment by Giwayume** [comment]
> The reason performance dies as you zoom in is simply because the shader is ran for more fragments.

This is not what I'm seeing, because the entire shape is already taking up the entire size of the game window, running at about 50fps, then as I zoom in more it dips into the single digits.

**Comment by SlugFiller** [comment]
@Giwayume Are you zooming near the center, or the edge? From what I saw, the shape you're using is a circle, which is implemented as 4 arcs (the editor shows 2, but the engine pre-divides arcs to 90 degrees, since they are easier to work with mathematically). So a possible culprit is that the number of subdivisions necessary before a given fragment is determined to be outside the arc's approximate bounding box increases as you zoom. From the get go, the arc's bounding box approximation isn't as good as for beziers. And a good approximation is necessary for early-stopping on the iterations.

**Comment by voidshine** [comment]
Yep, native vector graphics support for Godot could bring on Flash-like popularity for the engine, as vector unlocks many new kinds of game and application possibilities. How many times do we need to see grainy raster graphics before we accept that resolutions & devices will vary widely? The future is scalable. The present is, too. Even if the implementation just rasterized to the right size texture at runtime, that'd be good enough for many use cases.

**Comment by TechnoPorg** [comment]
How does this proposal tie in to the use of SVGs as `Texture2D`s, as compared to the current handling strategy, which is just to rasterize them on import? It would be nice to have adjustable vector graphics textures for things like shader parameters, theme icons, etc.

**Comment by Calinou** [comment]
> How does this proposal tie in to the use of SVGs as `Texture2D`s, as compared to the current handling strategy, which is just to rasterize them on import? It would be nice to have adjustable vector graphics textures for things like shader parameters, theme icons, etc.

This proposal is about rendering SVGs as polygons or SDFs, while your proposal is different entirely.

**Comment by SlugFiller** [comment]
@TechnoPorg Textures are necessarily raster. Any attempt to use SVG as a texture will therefore necessarily set it to a specific resolution. This can already be done directly today, by importing the SVG as a raster image. However, if you want the resolution to change in runtime, it is possible to do this by combining this suggestion with a viewport. i.e. rendering the vector to a viewport, and then using the viewport's texture as the texture. Since you still need to manually adjust the viewport's resolution, it is questionable how much of an advantage this gives.

It is impossible for an SVG to act as a "true" texture at infinite resolution. At most, it can be done as a shader. But that would make it difficult to combine with other shaders, and also, as demonstrated in the above shader-based vector graphics implementation, is incredibly slow, putting way more stress on the GPU than can be managed in a game that must run in real time.

**Comment by fire** [comment]
One of the thorvg team is employed at https://lottiefiles.com/ which takes after effect animated motion artwork and render it.

So both lottie (standard) and rive (proprietary) are approaches to render.

**Comment by TechnoPorg** [comment]
@SlugFiller Those are excellent points that I had not considered, and I like the idea of the viewport solution.
What I had originally been thinking of was some kind of `VectorTexture` type that renders the vector image, to be used as the texture. It would then cache that until its `transform` or `animate` property is changed, perhaps by script, or by some special `VectorSprite` node, which forwards its transform changes to its texture. At that point the texture re-rasterizes the vector image at the highest possible display resolution and displays the new texture.
That concept probably has all kinds of flaws, so please forgive my ignorance. If it could be made to work, though, I think something like a `VectorSprite` could be faster than calling a `draw_filled_curve` call every frame.

**Comment by SlugFiller** [comment]
@TechnoPorg If the vector doesn't change, you don't need to call `draw_filled_curve` every frame. Only every change. If it does change, it would have to be rasterized every frame, which would be just as slow. Also, a viewport set to UPDATE_ONCE would achieve the same effect of pre-rasterizing. It would only draw once, and can be used as a texture until it needs to be updated, at which point it can be set to UPDATE_ONCE again. Adding a special node or resource to achieve that wouldn't serve any purpose. It would be no more efficient, and the amount of things that need to be handled manually (e.g. deciding the resolution of the raster) wouldn't be reduced even by a bit. Nor would there be any performance boost to it (Whichever rasterizer is used, it's not going to beat an in-engine one. If it could, the same rasterization technique could be used by the engine. At the very least, a software rasterizer is not going to beat a hardware-aided one at anything other than trivial resolutions).

**Comment by SlugFiller** [comment]
@fire Now that I'm done working on this, I took the time to see how Tove2D implements its path to mesh. It's... "interesting". It uses a library called PolyPartition, and performs a few steps:
1. It "removes holes" by adding edges connecting "holes" (counter-clockwise polygons) with the main "shape" (clockwise polygon), creating a single connected shape that has 0-width "tunnels" connecting any holes to the "outside".
2. It uses ear-clipping to triangulate the shape. Ear clipping is known to not work if the shape has any holes or self-intersections. Holes were removed in the previous step, but intersecting edges could still make the algorithm fail at this point.
3. It uses Hertel-Mehlhorn to reduce the triangulation to convex polygons. This works by "merging" adjacent triangles into polygons while ensuring they remain convex.
4. It iterates over the convex polygons, and triangulates them by ear-clipping. You may be wondering what's the point of doing this after doing the previous step. There is no point. The person who wrote this is just stupid.

All in all, the algorithm is horribly inefficient, and can fail on intersecting edges, e.g. in a simple pentagram. Worse yet, PolyPartition actually contains a different algorithm called "Triangulate_MONO" which is basically Bently-Ottmann, but without intersection checking. It would still fail on self-intersection, but would run at a fraction of the time.

Incidentally, Godot already contains an ear-clipping triangulator, used by Polygon2D, so including a third party library for it would have been redundant.

**Comment by Giwayume** [comment]
I've found that a constrained delaunay implementation works fairly well for generating a triangulation with holes. The idea being, run the fill rule in the center of each triangle generated to find the holes.

**Comment by fire** [comment]
I didn't look into this deeply. @SlugFiller thanks for summary.

![image](https://github.com/godotengine/godot-proposals/assets/32321/23a64ded-dc9e-453b-a6ff-fc3c4c982b32)

Have a SVG to mesh cat :D. This was generated by the Tove2D PolyPartition workflow.

Hope to find a way to get that into Godot Engine properly.

**Comment by Sosasees** [comment]
> The person who wrote this is just stupid.

this is very rude, please remove this

**Comment by SlugFiller** [comment]
@Giwayume Delauney will not only ignore edge intersections, but the edges themselves, as it does not accept them as input. It can choose to split triangles using an edge that crosses one of the original edges, causing it to have an incorrect shape, regardless of which triangles you leave out.

@fire If my patch gets merged, you can use the methods I expose on `Geometry2D` to tessellate and triangulate the paths. It might be a bit more work, but performance and stability should be way better. Or you can patch Tove2D to use `Triangulate_MONO` instead of `Triangulate_HM`. The prior probably has a better chance of getting into Godot, because, less 3rd-party dependencies.

**Comment by Giwayume** [comment]
> @Giwayume Delauney will not only ignore edge intersections, but the edges themselves, as it does not accept them as input. It can choose to split triangles using an edge that crosses one of the original edges, causing it to have an incorrect shape, regardless of which triangles you leave out.

You need to look up what a "constrained" delaunay is, not a regular delaunay. From your description, it appears you don't understand what I'm talking about.

I have actually implemented this, it works.

**Comment by felaugmar** [comment]
Godot uses **ThorVG** for SVG drawing in 4.x.
ThorVG supports canvas oriented drawing (see **Shapes** in [Link)](https://www.thorvg.org/tutorial#comp-kp0z2k4o).

The API could follow ThorVG's API and exposed like below:

```gdscript
func _generate_texture(width: float, height: float, corner_radius_xy: float) -> Texture2D:
  # Generate a shape.
  var shape := CanvasShape()

  # Append a rounded rectangle to the shape (x, y, w, h, rx, ry).
  shape.append_rect(0, 0, width, height, corner_radius_xy, corner_radius_xy)

  # Set shape color.
  shape.fill(Color.DARK_GRAY)

  # Generate a canvas.
  var canvas := Canvas(width, height)

  # Push the shape into the canvas.
  canvas.push(shape)

  return canvas.get_texture()
```

![image](https://github.com/godotengine/godot-proposals/assets/11199582/fd4f9478-3650-4802-b8d7-12cec915b12e)

**Comment by felaugmar** [comment]
> very confusing if co-existing with the current API

The idea was to co-exist and deliver software rendering (allowing thread rendering), but I agree on getting confusing.

**Comment by fire** [comment]
My original design was for thorvg to first export to pixel buffers. Once things are futher along we can export static and animations as vector buffers for godot engine. 

We don’t expose the thorvg api bare.

**Comment by SlugFiller** [comment]
@felaugmar  The issue with that approach is that ThorVG, as it appears in Godot, uses a 100% software renderer. It's good enough for turning icons to bitmaps in the editor. It is not, however, sufficient for rendering vector animations, in real time, at any resolution expected of a modern game.

Even a meshing method, with GPU draw, is so slow that you can only afford 50-80 sprites on screen before going below 60 FPS. And that's running on an Intel i7. A 100% software scanline renderer would be even slower.

If a render-to-texture approach was to be used, something like Vello or at least Pathfinder, would be more appropriate, as it at least uses GPU acceleration to achieve 60 FPS on large scenes. They do carry other downsides, and would be a new 3rd party dependency if taken as-is.

One thing that could be a benefit _in combination_ with a `draw_` command is an SVG simplifier, like usvg. However, something like that can just as easily be implemented in GDScript and put up as an addon.

**Comment by sosasees** [comment]
since @SlugFiller's comment, it seems even more unlikely that real-time vector graphics rendering will be added to Godot.
until then, i would make my scalable game graphics from _distance field shaders_ and/or _icon fonts_.

**Comment by QbieShay** [comment]
To add to SlugFiller's comment, we discussed the relative PR and the general consensus from the maintainers is that this feature opens up a new usecase for Godot and we're not sure that it's possible, resource wise, to support this long-term, offer bugfixes and new features to the people that would start using Godot more like they'd use flash for making games before.
We do not want to ship features that we're not sure we can maintain.

Due to these concerns, the first step is to figure out if this can be made into a gdextension, and if not, what's missing in order to make it into a gdextension.

If the extension route is not possible, we'll evaluate whether this is a use-case we can afford to support or not.

My own personal take (so take this as a one person opinion and not the rendering team's final stance) is that it's better for Godot to focus on properly supporting the feature set that it already offers and improve on that, and leave vector rendering to either plugins or a specialized engine. Godot never set out to be a replacement for flash.

I know this comment will displease all of you who would like to have these features, but I think it's important to set the expectations right for everyone participating.

SlugFiller has asked for help in figuring out how to decouple the svg rendering from Godot's core rendering loop, so if anyone wants to help with that please do get in touch in rocket chat https://chat.godotengine.org/channel/rendering/

**Comment by HendrixString** [comment]
Hi @here
I am probably late to the party,
I was told that this vector engine I have written might interest you
https://github.com/micro-gl/micro-tess

**Comment by FedTheCat** [comment]
Rive just announced they are working to implement their runtime into Godot. Fingers crossed it gets in sooner rather than later!

https://twitter.com/guidorosso/status/1714737152633008500?t=Gf-8igKuoDWwUIXtEIEkHQ&s=19

**Comment by PluckyDuck99** [comment]
> Rive just announced they are working to implement their runtime into Godot. Fingers crossed it gets in sooner rather than later!
> 
> https://twitter.com/guidorosso/status/1714737152633008500?t=Gf-8igKuoDWwUIXtEIEkHQ&s=19

That will be amazing, but I think it will take some time. Godot doesn't allow you to integrate your own renderer yet :(

**Comment by SpudnikGames** [comment]
Has there been any word or update on this becoming a baseline possibility in Godot? Id like to use Vector Graphics directly in Godot and it looks like the last comments were a while ago. Anyone know if there are any plans for this in the 4.X branches of being baseline in some way or has that happened and I just missed it?

**Comment by fire** [comment]
See https://github.com/godotengine/godot/pull/91580 for a lottie work in progress I've been nudging.

**Comment by PluckyDuck99** [comment]
> See [godotengine/godot#91580](https://github.com/godotengine/godot/pull/91580) for a lottie work in progress I've been nudging.

Now the Rive renderer, which also renders vectors in real time is open source and can be integrated with Godot if someone work on it. It could be a good possibility to consider [Rive render](https://github.com/rive-app/rive-renderer)

**Comment by octanejohn** [comment]
thorvg is all ready in use plus is easier with lottie

**Comment by projectitis** [comment]
ThorVG is a rendering engine. Rive is an animation format (like Lottie but more powerful). They both work together - ThorVG can be used to render Lottie animations AND it can be used to render Rive animations.

I ported the Rive renderer to ThorVG some time ago. It worked perfectly - however, Rive evolved and gained many features that ThorVG didn't have yet, and my interest waned :) I may look again at some stage because ThorVG has been evolving steadily.

However, ThorVG renders to a bitmap canvas, and what we really need is an OpenGL rendering engine that generates meshes etc to take advantage of hardware. ThorVG did have the start of an OpenGL renderer, but not sure how far it got. There is a Rive implementation for Unity that works in this way.

**Comment by SlugFiller** [comment]
The people from Rive have actually shown interest in making a Godot port/integration/plugin. But it is not currently possible with the things Godot exposes, or rather, doesn't. Rive operates by using the stencil method for winding counting. However, while Godot does request the stencil buffer, it does not expose it to canvas items, and certainly not to plugins.

The other known methods for hardware rendering vectors is pathfinder, which operates as a compute shader, and would therefore require an ability to produce a texture from a compute shader; and mesh generation, which is still mostly CPU-bound, due to the need to compute path self-intersections.

By contrast, ThorVG, being a software renderer, has very high compatibility. Rendering doesn't need anything exposed other than the CPU, and transforming the output buffer to a texture doesn't require any odd GPU-GPU copies, or GPU-CPU-GPU round-trip transfers. The cost is, of course, performance. It's useful for static graphics, but is not good enough for creating a couple of hundreds of animated enemies on the screen at once. Even meshing, which performs the fill on the GPU, proved not performant enough for gaming.

**Comment by PluckyDuck99** [comment]
> The people from Rive have actually shown interest in making a Godot port/integration/plugin. But it is not currently possible with the things Godot exposes, or rather, doesn't. Rive operates by using the stencil method for winding counting. However, while Godot does request the stencil buffer, it does not expose it to canvas items, and certainly not to plugins.
> 
> The other known methods for hardware rendering vectors is pathfinder, which operates as a compute shader, and would therefore require an ability to produce a texture from a compute shader; and mesh generation, which is still mostly CPU-bound, due to the need to compute path self-intersections.
> 
> By contrast, ThorVG, being a software renderer, has very high compatibility. Rendering doesn't need anything exposed other than the CPU, and transforming the output buffer to a texture doesn't require any odd GPU-GPU copies, or GPU-CPU-GPU round-trip transfers. The cost is, of course, performance. It's useful for static graphics, but is not good enough for creating a couple of hundreds of animated enemies on the screen at once. Even meshing, which performs the fill on the GPU, proved not performant enough for gaming.

Interesting.. I thought it would be easier to integrate Rive with Godot. In a discussion, one of the Rive developers mentioned that they needed a few things to integrate the renderer with Godot and that everything seemed to be well underway and available for this...  [Rive mention](https://github.com/godotengine/godot-proposals/issues/4287#issuecomment-1812810331)

**Comment by SlugFiller** [comment]
If you look closely at what the Rive developer is saying, it's basically that they're missing any way to interface with Godot's graphics engine in a meaningful way. They are asking for direct access to the GL/Vulkan context, which is pretty low level, and not particularly likely to happen (It would be too backend specific, and not work very well for DX12 or Metal backends). It is a very far cry from "Well underway and available"

**Comment by Reuzi3** [comment]
There's not much progress yet, but I was thinking about the possibility of porting a library like [Dotlottie-rs](https://github.com/LottieFiles/dotlottie-rs) (in Rust) using GDExtension to make it a plugin. This could enable using .lottie SVG animations, and Dotlottie seems like a versatile and powerfull option. There's a PR in progress for a Texture2D that supports Lottie, but it wouldn't provide state machines or much flexibility. By porting Dotlottie we could gain more advantages. What would be the downsides of this, aside from the difficulty of porting it? I've seen many libraries brought to Godot using Godot-Rust.

**Comment by vybr** [comment]
I'm finding myself needing this more and more as my current projects are using vector assets for both the game world and UI. The current workflow of saving things in a raster format has a lot of downsides. 

As op mentioned, I think it's quite important that the vector assets can be created and manipulated at runtime as this makes it compatible with animations and tweens and enables interactivity. The ability to simply change a colour or gradient in a shape would unlock so much e.g. button hover animations or a clothing/equipment dye system which I'm currently having to implement with shaders.

Bonus points if supports simple effects, mainly drop shadows and inner shadows (with blurring).

In a perfect world I'd be able copy + paste SVGs from Figma/Affinity Designer straight into Godot... or even make them directly in the engine with a custom editor.

**Comment by Fz0011** [comment]
I still believe that Rive is the best way at the moment, but ThorVG and Lottie is evolving. Lottie  is a .json format for vector-based animation that is open source. It could be more flexible. Here, we have a demonstration of Lottie working using ThorVG to draw (which is already integrated into Godot) to render svgs. [ThorVG-Lottity](https://github.com/BenjaminHalko/Lottity/tree/main)

**Comment by Calinou** [comment]
@Fz0011 There's a proposal tracking Lottie support already: https://github.com/godotengine/godot-proposals/issues/9104

**Comment by Calinou** [comment]
> I see a lot of people talking about Rive and different vector-graphic formats, but unless I'm misunderstanding, couldn't we just adress the problem in this proposal with ThorVG? ThorVG is obviously capable of rasterizing svg's in realtime. You can currently implement your own rudimentary runtime-rasterizer by running svg_from_string() every frame. My decade-old laptop handles that without issue.

This is available in 4.5 as [SVGTexture](https://github.com/godotengine/godot/pull/105375) 🙂

**Comment by Calinou** [comment]
> For thorVG to perform well in production with vector animations, it would need GPU acceleration or a similar backend optimization something it currently lacks

We don't actually need a dedicated GPU backend for that. We just need to output a mesh from the vertex data retrieved from ThorVG. We do this already for TextMesh, so it can be done in 2D too.

Of course, it's more involved if you want to support gradients and antialiasing without needing 2D MSAA (feather gradients), but that's the basic idea.

**Comment by fire** [comment]
It's also theoretically possible to translate a lottie file into a godot 2d scene file.

**Comment by lostminds** [comment]
> We don't actually need a dedicated GPU backend for that. We just need to output a mesh from the vertex data retrieved from ThorVG, then render it in MeshInstance2D (or MultiMeshInstance2D). We do this already for TextMesh in 3D, so it can be done in 2D too. 

Being able to get mesh data from ThorVG to render that in the Godot renderer pipeline would be great. But if we're exploring tighter integration with ThorVG for drawing dynamic vector graphics in real time I think the svg generation and parsing required is currently a bigger bottleneck. So I'd say exposing some sort of way to directly define and interact with ThorVG shape entities to be rendered instead of needing to generate an svg code string and then have ThorVG parse that before rendering would be better to prioritize.

**Comment by SlugFiller** [comment]
ThorVG is the wrong tree to bark up. It would bare the same issues with the PR I made: It isn't performant enough for anything other than static UI elements. Despite carefully optimizing mesh generation for real time, I was only able to reach roughly 60 entities before the framerate started dropping. ThorVG's mesh generator is highly unlikely to be optimized for the same purpose.

Mesh generation is a process that doesn't parallelize very well, due to the need to examine path self-intersection, and handle holes in the path. And done serially in CPU, it can become a bottleneck.

Real time systems like Rive, skip mesh generation, and use stencil counting instead. Even with a lot of overdraw, rendering to a stencil is far more parallelizable, and is done completely in GPU.

The same could achieved in Godot directly if there was an ability to use the stencil buffer in 2D, to do the counting and create the mask. GPU tessellation could probably be hacked together using multimesh. However, designs for stencil buffer usage are currently only for 3D, and are in very early stage.

**Comment by hermet** [comment]
> ThorVG is the wrong tree to bark up. It would bare the same issues with the PR I made: It isn't performant enough for anything other than static UI elements. Despite carefully optimizing mesh generation for real time, I was only able to reach roughly 60 entities before the framerate started dropping. ThorVG's mesh generator is highly unlikely to be optimized for the same purpose.
> 
> Mesh generation is a process that doesn't parallelize very well, due to the need to examine path self-intersection, and handle holes in the path. And done serially in CPU, it can become a bottleneck.

@SlugFiller Hi, Please check this as well: https://thorvg-perf-test.vercel.app/

I'm curious about how complex the mesh being generated is. (if you have any screenshots probably, we are easier to understand the situation.) If the mesh generator process is a performance bottleneck, do you have any evidence or data could support that? If by 'mesh generator' you're referring to the tessellator stage, I would expect that it doesn't have a significant impact on performance than other parts like sw rasterization currently a way of working on godot side.

ThorVG already supports multi-threading though Godot is currently using a single thread, so there's room for improvement, obviously.

**Comment by SlugFiller** [comment]
> Please check this as well: https://thorvg-perf-test.vercel.app/

All of these examples contain only 2-3 vectors.

> if you have any screenshots

I've got better: I've got an example scene you can test [in the relevant PR page](https://github.com/godotengine/godot/pull/75278#issuecomment-1519193860). It can also render a single entity pretty quickly. But drops below 60 FPS as early as having 70 entities on screen.

The thing is, I compared this to native canvas 2D performance in current web browsers (Firefox, Chrome), and my optimized version outperformed the browsers for the same vectors in the same resolution.

This performance isn't really acceptable for a game: Most games won't be limited to one character. Having more than 70 characters on screen at the same time is actually pretty common.

> If by 'mesh generator' you're referring to the tessellator stage

Not the tessellator. Tessellation turns a curved path into jagged polyline. Mesh generator turns jagged polyline with possible self-intersections, to triangles with no overlap. And that's the bottleneck, because it can't be parallelized. The algorithm for it is serial. It has to go vertex by vertex, intersection by intersection.

The part you _can_ parallelize, is that you can process different paths at the same time. But each path is processed serially. And you're still using a lot of CPU for it, since this can't be offloaded to the GPU.

Rive's secret is skipping that second step. Instead, it creates a mesh with overlaps, with exactly one triangle per edge of the polyline. It then renders that mesh to the stencil buffer, counting overlaps. It then uses the overlap count as a mask for the final drawing.

There's also an alternative method, used by Pathfinder and Vello, which performs scanline tracing similar to a software renderer, but redesigned to work well in the GPU, using parallel sorting and parallel sum of prefixes algorithms.

**Comment by problematik** [comment]
Hopefully this adds to the conversation - here is one example of Rive running 105 high dpi vector characters at 120fps

https://x.com/guidorosso/status/1609401978811191299

**Comment by hermet** [comment]
> All of these examples contain only 2-3 vectors

@SlugFiller you can scale down on the desktop, and see the 100 complex animations running together.

**Comment by Samvkes** [comment]
> > Please check this as well: https://thorvg-perf-test.vercel.app/
> 
> All of these examples contain only 2-3 vectors.
> 
> > if you have any screenshots
> 
> I've got better: I've got an example scene you can test [in the relevant PR page](https://github.com/godotengine/godot/pull/75278#issuecomment-1519193860). It can also render a single entity pretty quickly. But drops below 60 FPS as early as having 70 entities on screen.
> 
> The thing is, I compared this to native canvas 2D performance in current web browsers (Firefox, Chrome), and my optimized version outperformed the browsers for the same vectors in the same resolution.
> 
> This performance isn't really acceptable for a game: Most games won't be limited to one character. Having more than 70 characters on screen at the same time is actually pretty common.
> 
> > If by 'mesh generator' you're referring to the tessellator stage
> 
> Not the tessellator. Tessellation turns a curved path into jagged polyline. Mesh generator turns jagged polyline with possible self-intersections, to triangles with no overlap. And that's the bottleneck, because it can't be parallelized. The algorithm for it is serial. It has to go vertex by vertex, intersection by intersection.
> 
> The part you _can_ parallelize, is that you can process different paths at the same time. But each path is processed serially. And you're still using a lot of CPU for it, since this can't be offloaded to the GPU.
> 
> Rive's secret is skipping that second step. Instead, it creates a mesh with overlaps, with exactly one triangle per edge of the polyline. It then renders that mesh to the stencil buffer, counting overlaps. It then uses the overlap count as a mask for the final drawing.
> 
> There's also an alternative method, used by Pathfinder and Vello, which performs scanline tracing similar to a software renderer, but redesigned to work well in the GPU, using parallel sorting and parallel sum of prefixes algorithms.

Are you saying that your Godot project outperformed your Js canvas implementation? I’m not an expert but as far as I know the browser svg rendering is considered plenty fast for realtime stuff. If Thor matches that performance I think that’s actually really good.

**Comment by Fz0011** [comment]
> > ThorVG is the wrong tree to bark up. It would bare the same issues with the PR I made: It isn't performant enough for anything other than static UI elements. Despite carefully optimizing mesh generation for real time, I was only able to reach roughly 60 entities before the framerate started dropping. ThorVG's mesh generator is highly unlikely to be optimized for the same purpose.
> > Mesh generation is a process that doesn't parallelize very well, due to the need to examine path self-intersection, and handle holes in the path. And done serially in CPU, it can become a bottleneck.
> 
> [@SlugFiller](https://github.com/SlugFiller) Hi, Please check this as well: https://thorvg-perf-test.vercel.app/
> 
> I'm curious about how complex the mesh being generated is. (if you have any screenshots probably, we are easier to understand the situation.) If the mesh generator process is a performance bottleneck, do you have any evidence or data could support that? If by 'mesh generator' you're referring to the tessellator stage, I would expect that it doesn't have a significant impact on performance than other parts like sw rasterization currently a way of working on godot side.
> 
> ThorVG already supports multi-threading though Godot is currently using a single thread, so there's room for improvement, obviously.

This performance is more than enough. Fingers crossed that something like this will make its way into Godot sooner or later. It would be a milestone for Lottie and for many in the Godot community, and it would remain totally optional. Godot isn’t Flash Player... but something like this opens up new horizons for those seeking the golden age when everything was crisp and fun, even if performance was lower back then. 😄

**Comment by piterecapzs** [comment]
lottie or rive either one would be a huge improvement. I really hate having to use mipmaps and seeing my textures look blurry

**Comment by Calinou** [comment]
> lottie or rive either one would be a huge improvement. I really hate having to use mipmaps and seeing my textures look blurry

Note that with https://github.com/godotengine/godot-proposals/issues/7411, mipmaps can be made to look sharper (at the cost of visible transitions when the mip level changes, e.g. due to zoom). https://github.com/godotengine/godot-proposals/issues/2450 can further help, although that proposal is written with 3D rendering in mind.

**Comment by maidopi-usagi** [comment]
I’m currently experimenting with integrating the Rive Renderer into Godot via GDExtension, and I’ve made some progress so far. It’s not really ready for use yet, as I’m still iterating on it for my own project, but it can serve as a proof of concept.
https://github.com/maidopi-usagi/RiveGD

<img src="https://github.com/user-attachments/assets/3e05f52d-6b5c-4b64-aa8d-ee32bb1d1284" width="100%" />

<img src="https://github.com/user-attachments/assets/60ab0db4-9f11-4778-8f70-89e278516135" width="100%" />
